File: split_pdf/workspace.go
// Workspace-relative path resolution against a shared WORKSPACE_ROOT.
// Duplicated from ../core/workspace.go — this repo's convention is each
// service carries its own copy of shared logic rather than a shared module.
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func filesDir(workspaceRoot, workspaceID string) string {
	return filepath.Join(workspaceRoot, workspaceID, "files")
}

// resolveWorkspacePath resolves relPath against {workspaceRoot}/{workspaceID}/files,
// rejecting absolute paths and any path (including via symlinks) that would
// escape that directory. If mustExist is true, the resolved path must
// already exist on disk.
func resolveWorkspacePath(workspaceRoot, workspaceID, relPath string, mustExist bool) (string, error) {
	base := filesDir(workspaceRoot, workspaceID)
	if _, err := os.Stat(base); err != nil {
		return "", fmt.Errorf("workspace not found: %s", workspaceID)
	}
	if filepath.IsAbs(relPath) {
		return "", fmt.Errorf("relative path must not be absolute: %s", relPath)
	}

	absBase, err := filepath.Abs(base)
	if err != nil {
		return "", fmt.Errorf("resolve workspace files dir: %w", err)
	}

	candidate := filepath.Join(base, relPath)
	absCandidate, err := filepath.Abs(candidate)
	if err != nil {
		return "", fmt.Errorf("resolve path: %w", err)
	}
	if absCandidate != absBase && !strings.HasPrefix(absCandidate, absBase+string(os.PathSeparator)) {
		return "", fmt.Errorf("relative path escapes workspace: %s", relPath)
	}

	realBase, err := filepath.EvalSymlinks(base)
	if err != nil {
		return "", fmt.Errorf("resolve workspace files dir: %w", err)
	}

	existingAncestor := absCandidate
	var remainder []string
	for {
		if _, err := os.Lstat(existingAncestor); err == nil {
			break
		}
		existingAncestor, remainder = filepath.Dir(existingAncestor), append([]string{filepath.Base(existingAncestor)}, remainder...)
		if existingAncestor == string(os.PathSeparator) || existingAncestor == "." {
			break
		}
	}
	realExisting, err := filepath.EvalSymlinks(existingAncestor)
	if err != nil {
		return "", fmt.Errorf("resolve path: %w", err)
	}
	if realExisting != realBase && !strings.HasPrefix(realExisting, realBase+string(os.PathSeparator)) {
		return "", fmt.Errorf("relative path escapes workspace: %s", relPath)
	}
	resolved := filepath.Join(append([]string{realExisting}, remainder...)...)

	if mustExist {
		if _, err := os.Stat(resolved); err != nil {
			return "", fmt.Errorf("path not found in workspace: %s", relPath)
		}
	}
	return resolved, nil
}

File: detect_boxes/workspace.py
"""Workspace-relative path resolution against a shared WORKSPACE_ROOT.

Every service in this repo that touches workspace files carries its own copy
of this module (no shared package between services — see README). Mirrors
`core/workspace.go`'s `resolveWorkspacePath` semantics: relative paths are
resolved against `{workspace_root}/{workspace_id}/files`, rejecting absolute
paths and anything that would escape that directory (directly or via a
symlink).
"""

from __future__ import annotations

import os
from pathlib import Path


def resolve_workspace_path(
    workspace_root: Path, workspace_id: str, relative_path: str, must_exist: bool = False
) -> Path:
    if os.path.isabs(relative_path):
        raise ValueError(f"relative path must not be absolute: {relative_path!r}")

    files_dir = workspace_root / workspace_id / "files"
    if not files_dir.is_dir():
        raise ValueError(f"workspace not found: {workspace_id!r}")

    real_base = files_dir.resolve(strict=True)
    candidate = (files_dir / relative_path).resolve(strict=False)
    if candidate != real_base and real_base not in candidate.parents:
        raise ValueError(f"relative path escapes workspace: {relative_path!r}")

    if must_exist and not candidate.exists():
        raise ValueError(f"path not found in workspace: {relative_path!r}")

    return candidate


def workspace_root_from_env() -> Path:
    value = os.environ.get("WORKSPACE_ROOT")
    if not value:
        raise RuntimeError("WORKSPACE_ROOT must be set")
    return Path(value)
